### PR TITLE
vhost-user-fs: Add support for vhost-user reconnection

### PR DESCRIPTION
**Pull request by sboeuf**
This PR adds reconnection support to the virtio-fs device. It does so by adding the reconnection thread, in order to handle a potential disconnection. It also factorizes the slave request handler into the reconnection thread, so that only one thread is used for both purposes. This allow for some important factorization and removal of code.
The last part of this PR adds the reconnection support to the slave request handler itself.

### CODE REVIEW

File: virtio-devices/src/vhost_user/mod.rs [hunk 1]
```diff
@@ -140,9 +140,10 @@ pub struct ReconnectEpollHandler {
     pub acked_protocol_features: u64,
     pub socket_path: String,
     pub server: bool,
+    pub slave_req_handler: Option<MasterReqHandler<S>>,
 }
```
Comment (zhangjaycee) [inline virtio-devices/src/vhost_user/mod.rs:144]: I think the name of `ReconnectEpollHandler` should be more generic, as it is now also handling the slave requests.

File: virtio-devices/src/vhost_user/mod.rs [hunk 2]
```diff
@@ -127,8 +126,9 @@ pub const DEFAULT_VIRTIO_FEATURES: u64 = 1 << VIRTIO_F_RING_INDIRECT_DESC
     | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
 
 const HUP_CONNECTION_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 1;
+const SLAVE_REQ_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 2;
 
-pub struct ReconnectEpollHandler {
+pub struct ReconnectEpollHandler<S: VhostUserMasterReqHandler> {
```
Comment (zhangjaycee) [inline virtio-devices/src/vhost_user/mod.rs:131]: I think the name of `ReconnectEpollHandler` should be more generic, as it is now also handling the slave requests.

File: virtio-devices/src/vhost_user/fs.rs [hunk 3]
```diff
@@ -452,35 +438,63 @@ impl VirtioDevice for Fs {
...
+            queues.clone(),
+            queue_evts.iter().map(|q| q.try_clone().unwrap()).collect(),
+            &interrupt_cb,
+            backend_acked_features,
+            &slave_req_handler,
+        )
+        .map_err(ActivateError::VhostUserFsSetup)?;
+
+        // Run a dedicated thread for handling potential reconnections with
+        // the backend.
```
Comment (zhangjaycee) [inline virtio-devices/src/vhost_user/fs.rs:459]: The comments need some change, as the thread is also for slave request handling.

File: virtio-devices/src/vhost_user/fs.rs [hunk 4]
```diff
@@ -356,18 +357,20 @@ impl Fs {
                 avail_features: acked_features,
                 acked_features: 0,
                 queue_sizes: vec![queue_size; num_queues],
-                paused_sync: Some(Arc::new(Barrier::new(2))),
+                paused_sync: Some(Arc::new(Barrier::new(3))),
```
Comment (zhangjaycee) [inline virtio-devices/src/vhost_user/fs.rs:360]: Maybe we need to change the 3 back to 2 in the last two commits of this PR, as we are merging the reconnect thread and the slave thread into one.

File: virtio-devices/src/vhost_user/mod.rs (hunk 2 above)
Comment (sboeuf) [reply virtio-devices/src/vhost_user/mod.rs:131]: I agree, I'll add a new commit to take care of this.

File: virtio-devices/src/vhost_user/fs.rs (hunk 3 above)
Comment (sboeuf) [reply virtio-devices/src/vhost_user/fs.rs:459]: Agreed!

File: virtio-devices/src/vhost_user/fs.rs (hunk 4 above)
Comment (sboeuf) [reply virtio-devices/src/vhost_user/fs.rs:360]: Definitely. I'll fix this as well.

**Comment by sboeuf** [comment]
@zhangjaycee I addressed your comments, please have a look.

File: virtio-devices/src/vhost_user/fs.rs [hunk 5]
```diff
@@ -452,35 +438,63 @@ impl VirtioDevice for Fs {
...
             slave_req_handler,
-        });
+        };
 
         let paused = self.common.paused.clone();
         let paused_sync = self.common.paused_sync.clone();
-        let mut epoll_threads = Vec::new();
+
         let virtio_vhost_fs_seccomp_filter =
             get_seccomp_filter(&self.seccomp_action, Thread::VirtioVhostFs)
```
Comment (zhangjaycee) [inline virtio-devices/src/vhost_user/fs.rs:480]: When performing test with virtiofsd, I found some problem here:

As this `VhostUserEpollHandler` is also used for handling reconnection, if we will keep the seccomp rules of `VirtioVhostFs`, we need to extent the rule list in `virtio-devices/src/seccomp_filters.rs`:
```rust
fn virtio_vhost_fs_thread_rules() -> Vec<SyscallRuleSet> {
    vec![
        // ....
        allow_syscall(libc::SYS_socket),
        allow_syscall(libc::SYS_nanosleep),
        allow_syscall(libc::SYS_connect),
        allow_syscall(libc::SYS_sendto),
    ]
}
``` 
The reconnection works fine after adding these 4 syscalls to the whitelist.

Comment (sboeuf) [reply virtio-devices/src/vhost_user/fs.rs:480]: Nice! Let me add these and we should be good to go :)

Comment (sboeuf) [reply virtio-devices/src/vhost_user/fs.rs:480]: Done. Please give this one more try, and if you confirm this is all good, we can go ahead and merge it :)

Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/fs.rs:480]: Just tried, works fine. 😊

Comment (sboeuf) [reply virtio-devices/src/vhost_user/fs.rs:480]: Thank you @zhangjaycee , I really appreciate you validating this :)

**Review by rbradford — APPROVED** [review]

**Pull request merged**